retrofit2/src/main/java/com/linecorp/armeria/client/retrofit2/ArmeriaRetrofitBuilder.java

```diff
@@ -77,91 +84,47 @@
     private static final String SLASH = "/";
 
     private final Retrofit.Builder retrofitBuilder;
-    private final ClientFactory clientFactory;
-    @Nullable
-    private String baseUrl;
+    private final ImmutableMap.Builder<String, EndpointGroup> endpointGroups = ImmutableMap.builder();
+
     private boolean streaming;
     private Executor callbackExecutor = CommonPools.blockingTaskExecutor();
-    private BiFunction<String, ? super ClientOptionsBuilder, ClientOptionsBuilder> configurator =
-            DEFAULT_CONFIGURATOR;
 
-    /**
-     * Creates a {@link ArmeriaRetrofitBuilder} with the default {@link ClientFactory}.
-     */
-    public ArmeriaRetrofitBuilder() {
-        this(ClientFactory.ofDefault());
-    }
+    private Function<? super HttpUrl, ? extends EndpointGroup> endpointGroupMapping =
+            url -> Endpoint.of(url.host(), url.port());
 
-    /**
-     * Creates a {@link ArmeriaRetrofitBuilder} with the specified {@link ClientFactory}.
-     */
-    public ArmeriaRetrofitBuilder(ClientFactory clientFactory) {
-        this.clientFactory = requireNonNull(clientFactory, "clientFactory");
-        retrofitBuilder = new Retrofit.Builder();
-    }
+    // FIXME(trustin): Do some caching.
+    private BiFunction<? super EndpointGroup, ? super HttpUrl, ? extends WebClient> webClientMapping =
+            (endpointGroup, url) -> WebClient.of(url.isHttps() ? SessionProtocol.HTTPS : SessionProtocol.HTTP,
+                                                 endpointGroup);
 
-    /**
-     * Sets the API base URL.
-     *
-     * @see Builder#baseUrl(String)
-     */
-    public ArmeriaRetrofitBuilder baseUrl(String baseUrl) {
-        return baseUrl(URI.create(requireNonNull(baseUrl, "baseUrl")));
+    ArmeriaRetrofitBuilder(HttpUrl baseUrl) {
+        retrofitBuilder = new Retrofit.Builder().baseUrl(baseUrl);
     }
 
-    /**
-     * Sets the API base URL.
-     *
-     * @see Builder#baseUrl(String)
-     */
-    public ArmeriaRetrofitBuilder baseUrl(URI baseUrl) {
-        requireNonNull(baseUrl, "baseUrl");
-        checkArgument(SessionProtocol.find(baseUrl.getScheme()) != null,
-                      "baseUrl must have an HTTP scheme: %s", baseUrl);
-        final String path = baseUrl.getPath();
-        if (!path.isEmpty() && !SLASH.equals(path.substring(path.length() - 1))) {
-            throw new IllegalArgumentException("baseUrl must end with /: " + baseUrl);
-        }
-        this.baseUrl = baseUrl.toString();
+    public ArmeriaRetrofitBuilder endpointGroup(
+            Function<? super HttpUrl, ? extends EndpointGroup> endpointGroupMapping) {
```

Comment:
I'm not quite understanding, why do we need to map URL to endpoint group instead of just accepting a single endpoint group used for the retrofit client, similar to the normal web client?

Reply:
Retrofit has an interesting annotation called `@Url` which allows a user send a request to an arbitrary URL, and a user should be able to choose an `EndpointGroup` in that case, too. See `ArmeriaCallFactoryTest.urlAnnotation*()` for examples.

Reply:
I think it's similar to webclient as well when issuing a request to absolute URL. I guess it would still send to the registered endpoint group and just change the host header. I don't know if we really need to allow `@Url` to resolve to an endpoint group - if the client has no endpoint group registered the URL can be resolved dynamically, if it does we just send to the registered endpoint group changing the host header. This seems simpler than using a mapping and I think preserves retrofit behavior well in the case of no endpoint group.

Reply:
Not sure I understood what you suggest. Could you elaborate?

Reply:
> if the client has no endpoint group registered, the URL can be resolved dynamically, if it does, we just send to the registered endpoint group changing the host header.

There can be more than one `EndpointGroup` involved here, so some kind of mapping is inevitable:

```java
EndpointGroup groupA = ...;
EndpointGroup groupB = ...;

ArmeriaRetrofit.builder("https://group-a/")
               .endpointGroup(url -> {
                   switch(url.host()) {
                       case "group-a": return groupA;
                       case "group-b": return groupB;
                       default: Endpoint.of(url.host(), url.port());
                   }
               })
               .build();

service.fullUrl("https://group-b/foo");
```

Reply:
I think I can simplify this at least for the case when no `@Url` annotation is used, like making it unnecessary to call `.endpointGroup()` at all.

Reply:
Alternatively, we could provide a `WebClient` wrapper which intercepts a request and translates a certain host to an `EndpointGroup`, instead of providing this functionality only in `armeria-retrofit`.

Reply:
I'm still finding that having multiple endpoint groups at this layer seems like overkill. URLs are basically host header / TLS but are somewhat independent from endpoint group - we resolve them to an endpoint when a user doesn't provide their own, but otherwise should just use the user specified endpoint independent of host. I think in common usage it will almost always be one endpoint group per client, and worse case a user can make multiple clients.

Also realized that it seems like if a user really needs host-aware endpoint selection, it could be fairly simply implemented as an endpoint selector that delegates to group selectors based on host?

Reply:
It should look somewhat better now. What do you think?

- A user can now specify the base `WebClient` when building `Retrofit` with `ArmeriaRetrofit`.
  - In most cases, a user can specify a customized `WebClient` here.
- Removed `ArmeriaRetrofitBuilder.endpointGroup()`
- Replaced `ArmeriaRetrofitBuilder.webClient()` with `nonBaseWebClientFactory()`.
  - The biggest difference from the previous version is that it is used only for non-base URLs, i.e. a user does not need to call this in most cases.
  - A user can still replace a certain host with an `EndpointGroup` to leverage client-side load-balancing for non-base URLs. (See the example in `nonBaseClientFactory()` Javadoc)